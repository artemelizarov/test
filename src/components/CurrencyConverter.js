import React, { Component } from "react";
import CurrencySelect from "./CurrencySelect";
import { DEFAULT_CURRENCY, DEFAULT_CURRENCY_2 } from "../common/constants";

class CurrencyConverter extends Component {
  componentDidMount() {
    this.props.changeCurrencyFrom(this.props.defaultCurrency);
    this.props.changeCurrencyTo(
      this.props.defaultCurrency === DEFAULT_CURRENCY
        ? DEFAULT_CURRENCY_2
        : DEFAULT_CURRENCY
    );
  }

  handleChange = event => {
    const { name, value } = event.target;
    if (name === "currencyFrom") {
      this.props.changeCurrencyFrom(value);
    } else if (name === "currencyTo") {
      this.props.changeCurrencyTo(value);
    }
  };

  handleAmountFromChange = event => {
    const value = +event.target.value;
    this.props.changeAmountFrom(value);
  };

  handleAmountToChange = event => {
    const value = +event.target.value;
    this.props.changeAmountTo(value);
  };

  render() {
    const {
      currencies,
      currencyTo,
      currencyFrom,
      amountFrom,
      amountTo
    } = this.props;

    return (
      <div className="CurrencyConverter">
        <h2>Currency converter</h2>
        <div>
          <CurrencySelect
            label="Currency from"
            name="currencyFrom"
            currencies={currencies}
            currency={currencyFrom}
            onChange={this.handleChange}
          />
          <CurrencySelect
            label="Currency to"
            name="currencyTo"
            currencies={currencies}
            currency={currencyTo}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <label>
            Amount from:
            <input
              name="amountFrom"
              type="number"
              value={amountFrom}
              onChange={this.handleAmountFromChange}
            />
          </label>

          <label>
            Amount to:
            <input
              name="amountTo"
              type="number"
              value={amountTo}
              onChange={this.handleAmountToChange}
            />
          </label>
        </div>
      </div>
    );
  }
}

export default CurrencyConverter;
