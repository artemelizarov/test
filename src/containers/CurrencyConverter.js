import { connect } from "react-redux";
import CurrencyConverter from "../components/CurrencyConverter";
import {
  changeCurrencyFrom,
  changeCurrencyTo,
  changeAmountFrom,
  changeAmountTo
} from "../redux/actions";

const mapStateToProps = state => {
  return {
    currencyFrom: state.currencyConverter.currencyFrom,
    currencyTo: state.currencyConverter.currencyTo,
    amountFrom: state.currencyConverter.amountFrom,
    amountTo: state.currencyConverter.amountTo,
    rate: state.currencyConverter.rate,
    currencies: state.currencies,
    defaultCurrency: state.defaultCurrency
  };
};

const mapDispatchToProps = dispatch => ({
  changeCurrencyFrom: currencyFrom =>
    dispatch(changeCurrencyFrom(currencyFrom)),
  changeCurrencyTo: currencyTo => dispatch(changeCurrencyTo(currencyTo)),
  changeAmountTo: amountTo => dispatch(changeAmountTo(amountTo)),
  changeAmountFrom: amountFrom => dispatch(changeAmountFrom(amountFrom))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CurrencyConverter);
