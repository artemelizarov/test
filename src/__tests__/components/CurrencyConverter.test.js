import CurrencyConverter from "../../components/CurrencyConverter";

describe("render CurrencyConverter component", () => {
  let wrapper,
    _changeCurrencyFrom,
    _changeCurrencyTo,
    _changeAmountFrom,
    _changeAmountTo;

  beforeAll(() => {
    _changeCurrencyFrom = jest.fn();
    _changeCurrencyTo = jest.fn();
    _changeAmountFrom = jest.fn();
    _changeAmountTo = jest.fn();

    wrapper = mount(
      <CurrencyConverter
        defaultCurrency="USD"
        currencies={[
          { code: "USD", name: "United States dollar" },
          { code: "EUR", name: "Euro" }
        ]}
        changeCurrencyFrom={_changeCurrencyFrom}
        changeCurrencyTo={_changeCurrencyTo}
        changeAmountFrom={_changeAmountFrom}
        changeAmountTo={_changeAmountTo}
      />
    );

    let input;
    input = wrapper.find("select[name='currencyFrom']").first();
    input.instance().value = "EUR";
    input.simulate("change");

    input = wrapper.find("select[name='currencyTo']").first();
    input.instance().value = "USD";
    input.simulate("change");

    input = wrapper.find("input[name='amountFrom']").first();
    input.instance().value = 10;
    input.simulate("change");

    input = wrapper.find("input[name='amountTo']").first();
    input.instance().value = 20;
    input.simulate("change");
  });

  test("CurrencyConverter render", () => {
    expect(wrapper.find(".CurrencyConverter").length).toBe(1);
  });

  test("_changeCurrencyFrom is called", () => {
    expect(_changeCurrencyFrom).toBeCalledWith("EUR");
  });

  test("_changeCurrencyTo is called", () => {
    expect(_changeCurrencyTo).toBeCalledWith("USD");
  });

  test("_changeAmountFrom is called", () => {
    expect(_changeAmountFrom).toBeCalledWith(10);
  });

  test("_changeAmountTo is called", () => {
    expect(_changeAmountTo).toBeCalledWith(20);
  });
});

describe("snapshot-test CurrencyConverter component", () => {
  test("Renders correct properties", () => {
    shallowExpect(
      <CurrencyConverter
        defaultCurrency="USD"
        currencies={[
          { code: "USD", name: "United States dollar" },
          { code: "EUR", name: "Euro" }
        ]}
        changeCurrencyFrom={jest.fn()}
        changeCurrencyTo={jest.fn()}
        changeAmountFrom={jest.fn()}
        changeAmountTo={jest.fn()}
      />
    ).toMatchSnapshot();
  });
});
