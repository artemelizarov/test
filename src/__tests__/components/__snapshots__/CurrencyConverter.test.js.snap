// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot-test CurrencyConverter component Renders correct properties 1`] = `
<div
  className="CurrencyConverter"
>
  <h2>
    Currency converter
  </h2>
  <div>
    <CurrencySelect
      currencies={
        Array [
          Object {
            "code": "USD",
            "name": "United States dollar",
          },
          Object {
            "code": "EUR",
            "name": "Euro",
          },
        ]
      }
      label="Currency from"
      name="currencyFrom"
      onChange={[Function]}
    />
    <CurrencySelect
      currencies={
        Array [
          Object {
            "code": "USD",
            "name": "United States dollar",
          },
          Object {
            "code": "EUR",
            "name": "Euro",
          },
        ]
      }
      label="Currency to"
      name="currencyTo"
      onChange={[Function]}
    />
  </div>
  <div>
    <label>
      Amount from:
      <input
        name="amountFrom"
        onChange={[Function]}
        type="number"
      />
    </label>
    <label>
      Amount to:
      <input
        name="amountTo"
        onChange={[Function]}
        type="number"
      />
    </label>
  </div>
</div>
`;
